File: AccelerateWatch/dsbuffer.c
/*  =========================================================================
    dsbuffer - fixed-length circular buffer for windowed signal processing

    Copyright (c) 2016, Yang LIU <gloolar@gmail.com>
    =========================================================================
*/

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <assert.h>

#include "dsbuffer.h"
#include "kissfft/kiss_fftr.h"


struct _dsbuffer_t {
    float *data;
    size_t size;
    size_t head; // position of first value
    void (*pusher)(dsbuffer_t *f, float new_value); // func for pushing value

    // for FFT
    bool fft_supported;
    kiss_fftr_cfg fft_cfg; // fft configuration

    // for FIR filter
    const float *fir_taps;
    size_t num_fir_taps;
    float (*fir_getter)(dsbuffer_t *buf); // func of getting filtered signal
};


// Check if x is power of 2
static bool is_power_of_2 (size_t x) {
   return x && !(x & (x - 1));
}


static void dsbuffer_push_fast_with_fft (dsbuffer_t *self, float new_value) {
    self->data[self->head] = new_value;
    self->data[self->head + self->size] = new_value;
    self->head = (++self->head) & (self->size-1);
}


static void dsbuffer_push_normal_with_fft (dsbuffer_t *self, float new_value) {
    self->data[self->head] = new_value;
    self->data[self->head + self->size] = new_value;
    if (++self->head == self->size)
        self->head = 0;
}


static void dsbuffer_push_fast (dsbuffer_t *self, float new_value) {
    self->data[self->head] = new_value;
    self->head = (++self->head) & (self->size-1);
}


static void dsbuffer_push_normal (dsbuffer_t *self, float new_value) {
    self->data[self->head] = new_value;
    if (++self->head == self->size)
        self->head = 0;
}


// Get latest FIR filter output (fast version)
static float dsbuffer_fir_get_fast (dsbuffer_t *self) {
    float fvalue = 0;
    int index = self->head, i;
    for (i = 0; i < self->num_fir_taps; ++i)
        fvalue += self->data[(--index) & (self->size - 1)] * self->fir_taps[i];
    return fvalue;
}


// Get latest FIR filter output (normal version)
static float dsbuffer_fir_get_normal (dsbuffer_t *self) {
    float fvalue = 0;
    int index = self->head, i;
    for (i = 0; i < self->num_fir_taps; ++i) {
        index = (index != 0) ? (index - 1) : (self->size - 1);
        fvalue += self->data[index] * self->fir_taps[i];
    }
    return fvalue;
}


static void dsbuffer_print (dsbuffer_t *self, bool fft_supported) {
    assert (self);
    printf ("\ndsbuffer: size: %zu, fft_supported: %s, head: %zu\n",
            self->size, fft_supported ? "true" : "false", self->head);
    printf("===============================================\n");
    for (size_t t = 0; t < self->size; t++) {
        printf ("%.3f", self->data[t]);
        if (t < self->size - 1) {
            printf(", ");
        }
    }

    if (fft_supported) {
        printf(" | ");
        for (size_t t = self->size; t < 2*self->size; t++) {
            printf ("%.3f", self->data[t]);
            if (t < 2*self->size - 1) {
                printf(", ");
            }
        }
    }
    printf("\n\n");
}


// ---------------------------------------------------------------------------


dsbuffer_t *dsbuffer_new (size_t size, bool fft_supported) {
    if (fft_supported && size % 2 == 1) {
        printf("ERROR: buffer size must be even for FFT.\n");
        return NULL;
    }

    dsbuffer_t *self = (dsbuffer_t *) malloc (sizeof (dsbuffer_t));
    assert (self);

    // Create signal buffer and initilize to zero
    size_t alloc_size = fft_supported ? size * 2 : size;
    self->data = (float *) calloc (alloc_size, sizeof (float));
    assert (self->data);

    self->size = size;
    self->head = 0;

    // If filter length equals power of 2, use the fast version, otherwise the
    // normal version
    if (is_power_of_2 (size))
        self->pusher = fft_supported ?
                       dsbuffer_push_fast_with_fft :
                       dsbuffer_push_fast;
    else
        self->pusher = fft_supported ?
                       dsbuffer_push_normal_with_fft :
                       dsbuffer_push_normal;

    self->fft_supported = fft_supported;
    if (fft_supported) {
        self->fft_cfg = kiss_fftr_alloc (size, 0, NULL, NULL);
        assert (self->fft_cfg);
    }
    else
        self->fft_cfg = NULL;

    self->fir_taps = NULL;
    self->num_fir_taps = 0;
    self->fir_getter = NULL;

    return self;
}


void dsbuffer_push (dsbuffer_t *self, float new_value) {
    assert (self);
    return self->pusher (self, new_value);
}


void dsbuffer_dump (dsbuffer_t *self, float *output) {
    assert (self);
    assert (output);
    if (self->fft_supported)
        memcpy (output, self->data + self->head, sizeof (float) * self->size);
    else {
        memcpy (output,
                self->data + self->head,
                sizeof (float) * (self->size - self->head));
        memcpy (output + self->size - self->head,
                self->data,
                sizeof (float) * self->head);
    }
}


void dsbuffer_fftr (dsbuffer_t *self, dsbuffer_complex *output) {
    assert (self);
    assert (output);
    kiss_fftr (self->fft_cfg, &self->data[self->head], (kiss_fft_cpx *)output);
}


void dsbuffer_fft_freq (dsbuffer_t *self, float fs, float *output) {
    assert (self);
    assert (output);
    assert (fs > 0);
    float interval = fs/self->size;
    output[0] = 0;
    for (size_t idx = 1; idx < self->size/2+1; idx++) {
        output[idx] = output[idx-1] + interval;
    }
}


void dsbuffer_setup_fir (dsbuffer_t *self, const float *fir_taps, size_t num_taps) {
    assert (self);
    assert (self->size >= num_taps);
    if (self->size > num_taps)
        printf ("WARNING: dsbuffer size is larger than number of FIR taps.");
    self->fir_taps = fir_taps;
    self->num_fir_taps = num_taps;

    if (is_power_of_2 (self->size))
        self->fir_getter = dsbuffer_fir_get_fast;
    else
        self->fir_getter = dsbuffer_fir_get_normal;
}


float dsbuffer_latest_fir_output (dsbuffer_t *self) {
    assert (self);
    assert (self->fir_taps);
    return self->fir_getter (self);
}


void dsbuffer_fir_filter (dsbuffer_t *self, float *output) {
    assert (self);
    assert (self->fir_taps);
    assert (output);

    // Convolution
    for (size_t ind_fsig = 0; ind_fsig < self->size; ind_fsig++) {
        float s = 0.0;
        for (size_t ind_tap = 0; ind_tap < self->num_fir_taps; ind_tap++) {
            if (ind_fsig < ind_tap)
                break;
            // s += self->fir_taps[ind_tap] * input_signal[ind_fsig-ind_tap];
            s += self->fir_taps[ind_tap] *
                 self->data[(self->head + ind_fsig - ind_tap) % self->size];
        }
        output[ind_fsig] = s;
    }
}


void dsbuffer_clear (dsbuffer_t *self) {
    assert (self);
    memset (self->data, 0, sizeof (float) *
                           (self->fft_supported ? (2*self->size) : self->size));
    self->head = 0;
}


void dsbuffer_free (dsbuffer_t **self_p) {
    assert (self_p);
    if (*self_p) {
        dsbuffer_t *self = *self_p;
        free (self->data);
        if (self->fft_cfg)
            free (self->fft_cfg);
        free (self);
        *self_p = NULL;
    }
}


void dsbuffer_test () {

    #include "fir_taps.ini"

    dsbuffer_t *buf = NULL;
    float *dumped = NULL;
    size_t size;
    bool fft_supported;

    // 1
    buf = dsbuffer_new (num_fir_taps, false);
    assert (buf);
    dsbuffer_setup_fir (buf, fir_taps, num_fir_taps);

    for (size_t t = 0; t < 10000000; t++) {
        float x = t * 1.0;
        dsbuffer_push (buf, x);
        dsbuffer_latest_fir_output (buf);
    }

    dsbuffer_free (&buf);

    // 2
    size = num_fir_taps;
    buf = dsbuffer_new (num_fir_taps, false);
    assert (buf);
    dsbuffer_setup_fir (buf, fir_taps, num_fir_taps);

    dumped = (float *) malloc (sizeof (float) * size);

    float signal1[] = {1, 4, 2, 5};
    for (size_t t = 0; t < 4; t++) {
        dsbuffer_push (buf, signal1[t]);
        dsbuffer_dump (buf, dumped);
        for (size_t i = 0; i < size; i++)
            printf ("dumped idx: %zu, value: %.3f\n", i, dumped[i]);
        printf ("FIR output: %.3f\n", dsbuffer_latest_fir_output (buf));
    }

    dsbuffer_print (buf, false);

    free (dumped);

    float *output = (float *) malloc (sizeof (float) * size);
    assert (output);
    dsbuffer_fir_filter (buf, output);
    for (size_t i = 0; i < num_fir_taps; i++)
        printf ("%.3f ", output[i]);
    printf ("\n");
    free (output);

    dsbuffer_free (&buf);

    // 3. FFT

    float signal2[] = {1, 4, 2, 5, 6, 7, -1, -8};
    size = sizeof (signal2) / sizeof (float);

    buf = dsbuffer_new (size, true);
    assert (buf);
    dsbuffer_print (buf, true);

    dumped = (float *) malloc (sizeof (float) * size);

    for (size_t t = 0; t < size; t++) {
        dsbuffer_push (buf, signal2[t]);
        dsbuffer_print (buf, true);
        // if (t == 3)
        //     dsbuffer_clear (buf);
        dsbuffer_dump (buf, dumped);
        for (size_t i = 0; i < size; i++)
            printf ("dumped idx: %zu, value: %.3f\n", i, dumped[i]);
    }

    free (dumped);

    dsbuffer_complex *fft_data =
        (dsbuffer_complex *) malloc (sizeof (dsbuffer_complex) * (size/2+1));
    dsbuffer_fftr (buf, fft_data);

    float *fft_freq = (float *) malloc (sizeof (float) * (size/2+1));
    dsbuffer_fft_freq (buf, 9.0, fft_freq);

    for (size_t idx = 0; idx < size/2+1; idx++) {
        printf("idx: %zu, freq: %.3f, real: %.3f, imag: %.3f\n",
               idx, fft_freq[idx], fft_data[idx].real, fft_data[idx].imag);
    }

    dsbuffer_free (&buf);


    // 4.
    size = 64;
    fft_supported = false;
    buf = dsbuffer_new (size, fft_supported);
    assert (buf);
    dumped = (float *) malloc (sizeof (float) * size);
    assert (dumped);

    for (size_t t = 0; t < 100000; t++) {
        float a = (int)(rand()*10000.0/RAND_MAX)/100.0;
        dsbuffer_push (buf, a);
        // dsbuffer_print (buf, fft_supported);
        dsbuffer_dump (buf, dumped);
        // for (size_t i = 0; i < size; i++)
        //     printf("%.3f, ", dumped[i]);
        // printf("\n");
        // printf("push: %.3f, get: %.3f\n", a, dumped[size-1]);
        assert (dumped[size-1] == a);
    }

    free (dumped);
    dsbuffer_free (&buf);

    // 5.
    size = 64;
    fft_supported = true;
    buf = dsbuffer_new (size, fft_supported);
    assert (buf);
    dumped = (float *) malloc (sizeof (float) * size);
    assert (dumped);

    for (size_t t = 0; t < 100000; t++) {
        float a = (int)(rand()*10000.0/RAND_MAX)/100.0;
        dsbuffer_push (buf, a);
        // dsbuffer_print (buf, fft_supported);
        dsbuffer_dump (buf, dumped);
        // for (size_t i = 0; i < size; i++)
        //     printf("%.3f, ", dumped[i]);
        // printf("\n");
        // printf("push: %.3f, get: %.3f\n", a, dumped[size-1]);
        assert (dumped[size-1] == a);
    }

    free (dumped);
    dsbuffer_free (&buf);


    // 6.
    size = 100;
    fft_supported = true;
    buf = dsbuffer_new (size, fft_supported);
    assert (buf);
    dumped = (float *) malloc (sizeof (float) * size);
    assert (dumped);

    for (size_t t = 0; t < 100000; t++) {
        float a = (int)(rand()*10000.0/RAND_MAX)/100.0;
        dsbuffer_push (buf, a);
        // dsbuffer_print (buf, fft_supported);
        dsbuffer_dump (buf, dumped);
        // for (size_t i = 0; i < size; i++)
        //     printf("%.3f, ", dumped[i]);
        // printf("\n");
        // printf("push: %.3f, get: %.3f\n", a, dumped[size-1]);
        assert (dumped[size-1] == a);
    }

    free (dumped);
    dsbuffer_free (&buf);

    // 7.
    size = 100;
    fft_supported = false;
    buf = dsbuffer_new (size, fft_supported);
    assert (buf);
    dumped = (float *) malloc (sizeof (float) * size);
    assert (dumped);

    for (size_t t = 0; t < 100000; t++) {
        float a = (int)(rand()*10000.0/RAND_MAX)/100.0;
        dsbuffer_push (buf, a);
        // dsbuffer_print (buf, fft_supported);
        dsbuffer_dump (buf, dumped);
        // for (size_t i = 0; i < size; i++)
        //     printf("%.3f, ", dumped[i]);
        // printf("\n");
        // printf("push: %.3f, get: %.3f\n", a, dumped[size-1]);
        assert (dumped[size-1] == a);
    }

    free (dumped);
    dsbuffer_free (&buf);


    printf ("OK\n");
}

File: AccelerateWatch/DSBuffer.swift
/// DSBuffer
///
/// Created by Yang Liu (gloolar@gmail.com) on 16/6/20.
/// Copyright © 2016年 Yang Liu. All rights reserved.


import Foundation


/// Fixed length signal buffer (Float type) which is suitable for storing and processing a windowed time series.
public class DSBuffer {
    
    private var buffer: OpaquePointer
    private var size: Int
    
    private var fftIsSupported: Bool
    private var fftData: [dsbuffer_complex]?
    private var fftIsUpdated: Bool?
    
    
    /// Constructor
    ///
    /// - parameter size: Buffer length
    /// - parameter fftIsSupported: Whether FFT will be performed on the buffer
    /// - returns: DSBuffer object
    ///
    /// *Tips*: If you do not need to perform FFT on the buffer, set fftIsSupperted to false would accelerate more. If you need to perform FFT, set buffer size to power of 2 would accelerate more.
    init(_ size: Int, fftIsSupported: Bool = true) {
        if (fftIsSupported && size % 2 == 1) {
            print("WARNING: size must be even for FFT. Reset size to: %d.", size+1)
            self.size = size + 1
        }
        else {
            self.size = size
        }
        self.buffer = dsbuffer_new(size, fftIsSupported)
        
        self.fftIsSupported = fftIsSupported
        if (fftIsSupported) {
            self.fftData = [dsbuffer_complex](repeating: dsbuffer_complex(real: 0.0, imag: 0.0), count: self.size/2+1)
            self.fftIsUpdated = false
        }
    }
    
    
    /// Destructor
    deinit {
        dsbuffer_free_unsafe(self.buffer)
    }
    
    
    /// Push new value to buffer
    ///
    /// - parameter value: New value to be added
    func push(_ value: Float) {
        dsbuffer_push(self.buffer, value)
        if (self.fftIsSupported) {
            self.fftIsUpdated = false
        }
    }
    
    
    /// Get Buffer size
    var bufferSize: Int {
        return self.size
    }
    
    
    /// Dump buffer as array
    var signals: [Float] {
        var dumpedSignal = [Float](repeating: 0.0, count: self.size)
        dsbuffer_dump(self.buffer, &dumpedSignal)
        return dumpedSignal
    }
    
    
    /// Reset buffer to be zero filled
    func clear() {
        dsbuffer_clear(self.buffer)
        if (self.fftIsSupported) {
            self.fftIsUpdated = false
        }
    }
    
    
    // MARK: Vector operations
    
    /// Add value to each buffer data
    func add(value: Float) -> [Float] {
        var result = [Float](repeating: 0.0, count: self.size)
        dsbuffer_add(self.buffer, value, &result)
        return result
    }
    
    
    /// Multiply each buffer data with value
    func multiply(value: Float) -> [Float] {
        var result = [Float](repeating: 0.0, count: self.size)
        dsbuffer_multiply(self.buffer, value, &result)
        return result
    }
    
    
    /// Mean value
    var mean: Float {
        return dsbuffer_mean(self.buffer)
    }
    
    
    /// Vector length
    var length: Float {
        return dsbuffer_length(self.buffer)
    }
    
    
    /// Remove mean value (centralize)
    func centralized() -> [Float] {
        var result = [Float](repeating: 0.0, count: self.size)
        dsbuffer_remove_mean(self.buffer, &result)
        return result
    }
    
    
    /// Normalize vector to have unit length
    func normalizedToUnitLength(centralized: Bool) -> [Float] {
        var result = [Float](repeating: 0.0, count: self.size)
        dsbuffer_normalize_to_unit_length(self.buffer, centralized, &result)
        return result
    }
    
    
    /// Perform dot production with array
    func dotProduct(with: [Float]) -> Float {
        assert(self.size == with.count)
        return dsbuffer_dot_product(self.buffer, with)
    }
    
    
    // MARK: FFT
    
    // Perform FFT if it is not updated
    private func updateFFT() {
        assert (self.fftIsSupported)
        if (!self.fftIsUpdated!) {
            _ = fft()
        }
    }
    
    
    /// Perform FFT
    ///
    /// **Note for FFT related methods**:
    ///
    /// - Set fftIsSupported to true when creating the buffer.
    /// - Buffer size should be even. If you pass odd size when creating the buffer, it is automatically increased by 1.
    /// - Only results in nfft/2+1 complex frequency bins from DC to Nyquist are returned.
    func fft() -> (real: [Float], imaginary: [Float]) {
        assert (self.fftIsSupported)
        dsbuffer_fftr(self.buffer, &self.fftData!)
        self.fftIsUpdated = true
        return (self.fftData!.map{$0.real}, self.fftData!.map{$0.imag})
    }
    
    
    /// FFT sample frequencies
    ///
    /// - returns: array of size nfft/2+1
    func fftFrequencies(fs: Float) -> [Float] {
        assert (self.fftIsSupported)
        var fftFreq = [Float](repeating: 0.0, count: self.size/2+1)
        dsbuffer_fft_freq(self.buffer, fs, &fftFreq)
        return fftFreq
    }
    
    
    /// FFT magnitudes, i.e. abs(fft())
    ///
    /// - returns: array of size nfft/2+1
    func fftMagnitudes() -> [Float] {
        updateFFT()
        return self.fftData!.map{sqrt($0.real*$0.real + $0.imag*$0.imag)}
    }
    
    
    /// Square of FFT magnitudes, i.e. (abs(fft()))^2
    ///
    /// - returns: array of size nfft/2+1
    func squaredPowerSpectrum() -> [Float] {
        updateFFT()
        var sps = self.fftData!.map{($0.real*$0.real + $0.imag*$0.imag) * 2}
        sps[0] /= 2.0 // DC
        return sps
    }
    
    
    /// Mean-squared power spectrum, i.e. (abs(fft()))^2 / N
    ///
    /// - returns: array of size nfft/2+1
    func meanSquaredPowerSpectrum() -> [Float] {
        updateFFT()
        var pxx = self.fftData!.map{($0.real*$0.real + $0.imag*$0.imag) * 2 / Float(self.size)}
        pxx[0] /= 2.0 // DC
        return pxx
    }
    
    
    /// Power spectral density (PSD), i.e. (abs(fft()))^2 / (fs*N)
    /// 
    /// - returns: array of size nfft/2+1
    func powerSpectralDensity(fs: Float) -> [Float] {
        updateFFT()
        var psd = self.fftData!.map{($0.real*$0.real + $0.imag*$0.imag) * 2.0 / (fs * Float(self.size))}
        psd[0] /= 2.0 // DC
        return psd
    }
    
    
    /// Average power over specific frequency band, i.e. mean(abs(fft(from...to))^2)
    func averageBandPower(fromFreq: Float = 0, toFreq: Float, fs: Float) -> Float {
        assert (fromFreq >= 0)
        assert (toFreq <= fs/2.0)
        assert (fromFreq <= toFreq)
        
        updateFFT()
        
        // Compute index range corresponding to given frequency band
        // f = idx*df = idx*fs/N ==> idx = N*f/fs
        let fromIdx = Int(floor(fromFreq * Float(self.size) / fs))
        let toIdx = Int(ceil(toFreq * Float(self.size) / fs))
        
//        let fftBand = Array(self.fftData![fromIdx...toIdx])
        let bandPower = self.fftData![fromIdx...toIdx].map{$0.real*$0.real+$0.imag*$0.imag}
        
        // Averaging
        return bandPower.reduce(0.0, combine: +) / Float(toIdx - fromIdx + 1)
    }
    
    
    // MARK: FIR filter
    
    // Setup FIR filter
    func setupFIRFilter(FIRTaps: [Float]) {
        assert (self.size >= FIRTaps.count)
        dsbuffer_setup_fir(self.buffer, FIRTaps, FIRTaps.count)
    }
    
    
    /// Get latest FIR output
    func latestFIROutput() -> Float {
        return dsbuffer_latest_fir_output(self.buffer)
    }
    
    
    /// FIR filtered buffer
    func FIRFiltered() -> [Float] {
        var output = [Float](repeating: 0.0, count: self.size)
        dsbuffer_fir_filter(self.buffer, &output)
        return output
    }
    
    
    // Self test
    class func test() {
        print("DSBuffer test:\n\n")
        
        let size = 16
        let buf = DSBuffer(size, fftIsSupported: true)
        
        let signalData: [Float] = [1.0, 4, 2, 5, 6, 7, -1, -8]
        for value in signalData {
            buf.push(value)
//            print(buf.signals)
        }
        
        let fft = buf.fft()
        print(fft)
        
        let fftSM = buf.squaredPowerSpectrum()
        print(fftSM)
        
        
        buf.clear()
        for _ in 0 ..< 10000 {
            let a = Float(arc4random_uniform(100))
            buf.push(a)
            let signals = buf.signals
//            print(a)
//            print(signals)
            assert(signals[size-1] == a)
        }
        
        
        let norm = buf.normalizedToUnitLength(centralized: true)
        let coeff = vDotProduct(norm, v2: norm)
        print("coeff: %.2f\n", coeff)
        
        print("\nDSBuffer test OK.\n\n")
    }
    
}
